nexus/db-queries/src/db/queries/instance.rs

```diff
@@ -339,52 +457,251 @@ impl QueryFragment<Pg> for InstanceAndVmmUpdate {
-        out.push_sql("SELECT instance_found.");
-        out.push_identifier(instance_dsl::id::NAME)?;
-        out.push_sql(" AS found, instance_updated.");
-        out.push_identifier(instance_dsl::id::NAME)?;
-        out.push_sql(" AS updated");
-        out.push_sql(
-            " FROM instance_found LEFT JOIN instance_updated ON instance_found.",
-        );
-        out.push_identifier(instance_dsl::id::NAME)?;
-        out.push_sql(" = instance_updated.");
-        out.push_identifier(instance_dsl::id::NAME)?;
-        out.push_sql(")");
-
-        if self.migration.is_some() {
-            out.push_sql(", ");
-            out.push_sql("migration_result AS (");
-            out.push_sql("SELECT migration_found.");
-            out.push_identifier(migration_dsl::id::NAME)?;
-            out.push_sql(" AS found, migration_updated.");
-            out.push_identifier(migration_dsl::id::NAME)?;
-            out.push_sql(" AS updated");
-            out.push_sql(
-                " FROM migration_found LEFT JOIN migration_updated ON migration_found.",
-            );
-            out.push_identifier(migration_dsl::id::NAME)?;
-            out.push_sql(" = migration_updated.");
-            out.push_identifier(migration_dsl::id::NAME)?;
-            out.push_sql(")");
+        out.push_sql(") ");
+
+        fn push_select_from_result(
+            update: Option<&Update>,
+            out: &mut AstPass<'_, '_, Pg>,
+        ) {
+            if let Some(update) = update {
+                out.push_sql(update.name);
+                out.push_sql("_result.found, ");
+                out.push_sql(update.name);
+                out.push_sql("_result.updated");
+            } else {
+                out.push_sql("NULL, NULL")
+            }
         }
-        out.push_sql(" ");
 
         out.push_sql("SELECT vmm_result.found, vmm_result.updated, ");
-        out.push_sql("instance_result.found, instance_result.updated, ");
-        if self.migration.is_some() {
-            out.push_sql("migration_result.found, migration_result.updated ");
-        } else {
-            out.push_sql("NULL, NULL ");
+        push_select_from_result(self.instance.as_ref(), &mut out);
+        out.push_sql(", ");
+        push_select_from_result(self.migration_in.as_ref(), &mut out);
+        out.push_sql(", ");
+        push_select_from_result(self.migration_out.as_ref(), &mut out);
+        out.push_sql(" ");
+
+        out.push_sql("FROM vmm_result");
+        if self.instance.is_some() {
+            out.push_sql(", instance_result");
         }
-        out.push_sql("FROM vmm_result, instance_result");
-        if self.migration.is_some() {
-            out.push_sql(", migration_result");
+        if self.migration_in.is_some() {
+            out.push_sql(", migration_in_result");
+        }
+
+        if self.migration_out.is_some() {
+            out.push_sql(", migration_out_result");
         }
 
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod test {
```

Comment:
I figured that, since we were generating a huge pile of SQL here, it was worth having some tests to make sure the SQL we were generating was reasonable, and see when it has changed. These are based on the tests @smklein added for the virtual provisioning queries in https://github.com/oxidecomputer/omicron/pull/5089.